File: apps/api/internal/domain/invitation/invitation_entity.go
package invitation

import (
	"crypto/rand"
	"crypto/subtle"
	"encoding/hex"
	"errors"
	"time"
)

// InvitationStatus represents the status of an invitation.
type InvitationStatus string

const (
	InvitationStatusPending   InvitationStatus = "pending"
	InvitationStatusApproved InvitationStatus = "approved"
	InvitationStatusRejected InvitationStatus = "rejected"
	InvitationStatusExpired  InvitationStatus = "expired"
)

// InvitationStatusPtr returns a pointer to the given InvitationStatus.
func InvitationStatusPtr(s InvitationStatus) *InvitationStatus {
	return &s
}

// Invitation represents an invitation to join an organization.
type Invitation struct {
	ID             string
	OrganizationID string
	Email         string
	Role          InvitationRole
	Status        InvitationStatus
	Token         string
	InviterNotes  *string
	InviteeNotes  *string
	InvitedBy     string
	InvitedAt     time.Time
	RespondedAt   *time.Time
	ResponderID   *string
	ExpiresAt     time.Time

	// Populated fields (joined from other tables)
	OrganizationName string
	InviterName     string
}

// InvitationRole represents the role that will be assigned when an invitation is accepted.
type InvitationRole string

const (
	InvitationRoleAdmin    InvitationRole = "admin"
	InvitationRoleOperator InvitationRole = "operator"
	InvitationRoleViewer  InvitationRole = "viewer"
)

// IsValid returns true if the invitation role is valid.
func (r InvitationRole) IsValid() bool {
	switch r {
	case InvitationRoleAdmin, InvitationRoleOperator, InvitationRoleViewer:
		return true
	default:
		return false
	}
}

// ToOrgRole converts an invitation role to an organization role.
func (r InvitationRole) ToOrgRole() string {
	switch r {
	case InvitationRoleAdmin:
		return "admin"
	case InvitationRoleOperator:
		return "operator"
	case InvitationRoleViewer:
		return "viewer"
	default:
		return "viewer"
	}
}

// IsExpired returns true if the invitation has expired.
func (i *Invitation) IsExpired() bool {
	return time.Now().After(i.ExpiresAt)
}

// IsPending returns true if the invitation is pending.
func (i *Invitation) IsPending() bool {
	return i.Status == InvitationStatusPending
}

// CanBeAccepted returns true if the invitation can be accepted.
func (i *Invitation) CanBeAccepted() bool {
	return i.Status == InvitationStatusPending && !i.IsExpired()
}

// CreateInvitationRequest represents a request to create an invitation.
type CreateInvitationRequest struct {
	OrganizationID string
	Email         string
	Role          InvitationRole
	InviterNotes  string
}

// Validate validates the create invitation request.
func (r *CreateInvitationRequest) Validate() error {
	if r.OrganizationID == "" {
		return errors.New("organization ID is required")
	}
	if r.Email == "" {
		return errors.New("email is required")
	}
	if !r.Role.IsValid() {
		return errors.New("invalid role")
	}
	return nil
}

// AcceptInvitationRequest represents a request to accept an invitation.
type AcceptInvitationRequest struct {
	Token       string
	InviteeNotes string
}

// RespondInvitationRequest represents a request to respond to an invitation.
type RespondInvitationRequest struct {
	Token       string
	Notes       string
}


// GenerateSecureToken generates a secure random token for invitation links.
func GenerateSecureToken(length int) (string, error) {
	bytes := make([]byte, length)
	if _, err := rand.Read(bytes); err != nil {
		return "", err
	}
	return hex.EncodeToString(bytes), nil
}

// ConstantTimeCompare performs a constant-time comparison of two strings.
func ConstantTimeCompare(a, b string) bool {
	return subtle.ConstantTimeCompare([]byte(a), []byte(b)) == 1
}

File: packages/API_Client/src/domain/invitation/invitation-entity.ts
/**
 * Invitation domain types aligned with server models.
 */

import type { OrganizationRole } from "../organization/organization-entity";

// Invitation lifecycle states
export type InvitationLifecycle = "pending" | "approved" | "rejected" | "expired";

// Default invitation expiry in hours
export const DEFAULT_INVITATION_EXPIRY_HOURS = 168; // 7 days

/**
 * Invitation entity.
 */
export interface Invitation {
  id: string;
  organizationId: string;
  email: string;
  role: OrganizationRole;
  lifecycle: InvitationLifecycle;
  invitedBy: string;
  invitedAt: Date;
  expiresAt: Date;
  inviterNotes?: string;
  // Populated fields
  organizationName?: string;
  invitedByName?: string;
}

/**
 * Invitation list item (lightweight).
 */
export interface InvitationListItem {
  id: string;
  organizationId: string;
  organizationName: string;
  email: string;
  role: OrganizationRole;
  lifecycle: InvitationLifecycle;
  invitedAt: Date;
  expiresAt: Date;
}

/**
 * Create invitation request.
 */
export interface CreateInvitationRequest {
  organizationId: string;
  email: string;
  role: OrganizationRole;
  inviterNotes?: string;
}

/**
 * Invitation response request (accept/reject).
 */
export interface InvitationResponseRequest {
  notes?: string;
}

// API response types (raw from server)
export interface InvitationApiResponse {
  id: string;
  organization_id: string;
  email: string;
  role: string;
  lifecycle: string;
  invited_by: string;
  invited_at: string;
  expires_at: string;
  inviter_notes?: string;
  organization_name?: string;
  invited_by_name?: string;
}

/**
 * Check if invitation is pending.
 */
export function isInvitationPending(invitation: Invitation): boolean {
  return invitation.lifecycle === "pending";
}

/**
 * Check if invitation is expired.
 */
export function isInvitationExpired(invitation: Invitation): boolean {
  return new Date() > invitation.expiresAt;
}

/**
 * Check if invitation can be responded to.
 */
export function canRespondToInvitation(invitation: Invitation): boolean {
  return isInvitationPending(invitation) && !isInvitationExpired(invitation);
}
